File: .claude/hooks/leak-guard.py
#!/usr/bin/env python3
"""Leak-guard PreToolUse hook (issue #173): block user-local filesystem paths from
entering SHARED artifacts at write time, instead of relying on per-skill prose to
remember to grep (the failure mode that shipped a vault path to main — see #158).

Wired on Write/Edit/MultiEdit. Reads the Claude Code PreToolUse JSON envelope from
stdin, decides whether the target is a shared-artifact doc surface, scans the text
being written for the #158 leak class, and DENIES the call (exit 0 + a deny decision
on stdout) when a real leak is present. A clean write is allowed silently.

LEAK SET (a committed home/local/vault path — the #158 regression class):
  - /Users/<name>/...        absolute macOS home path
  - a leading ~/             home-relative path (word-boundary, not mid-token)
  - ~/.claude, ~/.usirin, ~/.agent   agent/tool home dirs
  - ~/code/...               home-dir sibling-repo clones
  - /vault/ , vault-style sibling paths
These mirror the patterns the review-doc gate greps for (review-doc SKILL.md section 4)
and the report skill's footer-privacy contract — this hook is their write-time
enforcement.

ALLOWLIST (legitimate content that must NOT be flagged):
  - Repo-relative paths (apps/web/..., .claude/skills/..., packages/...) — never absolute.
  - Sanctioned ephemeral scratch: /tmp/... (e.g. /tmp/write-code-*.md,
    /tmp/review-*-verdict-*.md). /tmp is a leak only if it is a /Users//~ path; a bare
    /tmp path encodes no user identity or local layout, so it is allowed.
  - Documented product paths like ~/.config/... (the kampus CLI credential store) — a
    ~/.config reference is a documented runtime location, not a leak of *this* machine.
  - Path-hygiene docs that spell the pattern out as a *pattern* rather than commit it as
    a real path: this hook file, the review-doc/triage/report skills, and any file whose
    own subject is path hygiene. Such files are exempt by path (DOC_SELF_EXEMPT) because
    they must name the very tokens they forbid (review-doc SKILL.md section 4 carve-out).

SHARED-ARTIFACT SURFACES (where the guard fires): committed doc surfaces only —
*.md anywhere, plus .decisions/ and .patterns/. Arbitrary source (a TS string literal
containing /Users in a test fixture) is out of scope, matching the AC's "scope the match
to shared-artifact doc surfaces, not arbitrary source."

A PreToolUse hook cannot see gh issue/PR bodies (those are Bash invocations, not Write);
the review-doc gate + report footer contract remain the net for posted bodies. This hook
covers the committed-file half of "shared artifacts", which is the #158 leak class.
"""

import json
import re
import sys

# --- shared-artifact doc surfaces -------------------------------------------------
DOC_SUFFIXES = (".md", ".mdx", ".markdown")
DOC_DIRS = ("/.decisions/", "/.patterns/")

# Files whose subject IS path hygiene: they must spell the forbidden tokens out as
# patterns. Matched as a path suffix so it works for absolute or repo-relative targets.
DOC_SELF_EXEMPT = (
	"/.claude/hooks/leak-guard.py",
	"/.claude/skills/review-doc/SKILL.md",
	"/.claude/skills/triage/SKILL.md",
	"/.claude/skills/report/SKILL.md",
	"/.claude/skills/report/footer.sh",
)

# --- the leak patterns ------------------------------------------------------------
# Each entry: (compiled regex, human reason). Order = report order.
LEAK_PATTERNS = [
	(re.compile(r"/Users/[A-Za-z0-9._-]+"), "absolute macOS home path (/Users/<name>/...)"),
	(re.compile(r"(?<![\w.])~/\.(claude|usirin|agent)\b"), "agent/tool home dir (~/.claude, ~/.usirin, ~/.agent)"),
	(re.compile(r"(?<![\w.])~/code/"), "home-dir sibling-repo clone (~/code/...)"),
	(re.compile(r"(?<![\w/])/vault/"), "vault path (/vault/...)"),
	# A leading ~/ home path in general — but NOT ~/.config (documented product path,
	# e.g. the kampus CLI credential store) and NOT the dirs already named above.
	(re.compile(r"(?<![\w.])~/(?!\.config\b)"), "home-relative path (~/...)"),
]

# Allow a bare /tmp/ path: it carries no user identity or local layout. The /Users
# pattern above already never matches /tmp, so no extra carve-out is needed here —
# this comment documents the deliberate non-match.


def is_shared_artifact(path: str) -> bool:
	p = path.replace("\\", "/")
	if any(p.endswith(s) for s in DOC_SUFFIXES):
		return True
	return any(d in p for d in DOC_DIRS)


def is_self_exempt(path: str) -> bool:
	# Normalize to a leading slash so a relative `.claude/...` target matches the same
	# suffix as an absolute `/…/.claude/...` one (Write passes absolute, but be robust).
	p = "/" + path.replace("\\", "/").lstrip("/")
	return any(p.endswith(s) for s in DOC_SELF_EXEMPT)


def extract_target_text(tool_name: str, tool_input: dict):
	"""Return (file_path, text-being-written) for the supported tools, else (None, '')."""
	file_path = tool_input.get("file_path", "") or ""
	if tool_name == "Write":
		return file_path, tool_input.get("content", "") or ""
	if tool_name == "Edit":
		return file_path, tool_input.get("new_string", "") or ""
	if tool_name == "MultiEdit":
		edits = tool_input.get("edits", []) or []
		return file_path, "\n".join(str(e.get("new_string", "")) for e in edits)
	return None, ""


def find_leaks(text: str):
	"""Yield (matched_substring, reason) for every leak in text."""
	for pat, reason in LEAK_PATTERNS:
		for m in pat.finditer(text):
			yield m.group(0), reason


def deny(reason_lines):
	body = "Leak-guard blocked this write (issue #173): a user-local path may not enter a shared artifact.\n\n"
	body += "\n".join(reason_lines)
	body += (
		"\n\nUse a repo-relative path instead (apps/web/..., .claude/skills/...). "
		"If this is genuinely a documented pattern, not a real path, the surface may "
		"need to be added to DOC_SELF_EXEMPT in .claude/hooks/leak-guard.py."
	)
	out = {
		"hookSpecificOutput": {
			"hookEventName": "PreToolUse",
			"permissionDecision": "deny",
			"permissionDecisionReason": body,
		}
	}
	print(json.dumps(out))
	sys.exit(0)


def main():
	try:
		payload = json.load(sys.stdin)
	except (json.JSONDecodeError, ValueError):
		sys.exit(0)  # malformed envelope: never block on parse failure

	tool_name = payload.get("tool_name", "")
	tool_input = payload.get("tool_input", {}) or {}
	file_path, text = extract_target_text(tool_name, tool_input)

	if not file_path or not text:
		sys.exit(0)
	if not is_shared_artifact(file_path):
		sys.exit(0)
	if is_self_exempt(file_path):
		sys.exit(0)

	leaks = list(find_leaks(text))
	if not leaks:
		sys.exit(0)

	seen = set()
	lines = []
	for matched, reason in leaks:
		key = (matched, reason)
		if key in seen:
			continue
		seen.add(key)
		lines.append(f"  - `{matched}` — {reason}")
	deny(lines)


if __name__ == "__main__":
	main()

File: packages/pipeline-cli/src/tools/leak-guard/leak-guard.ts
/**
 * `leak-guard` core — the pure, IO-free matcher that decides whether a write puts
 * a user-local filesystem path into a SHARED artifact (issue #173), moved into the
 * pipeline-cli registry (epic #994, Phase 2 / #999).
 *
 * This is the write-time enforcement of the repo's no-local-paths rule that used
 * to live only in per-skill prose (the failure mode that shipped a vault path to
 * main — see #158). `findLeaks(filePath, text)` returns every leak in `text` when
 * `filePath` is a shared-artifact surface and not self-exempt; otherwise an empty
 * list. Still no regex over arbitrary source — the match is scoped to the surfaces
 * `surfaceOf` names, nothing wider.
 *
 * There are **two** such surfaces, `"doc"` and `"shell"`. Shell was added because the
 * doc-only scoping's premise expired: it was written when all pipeline shell lived
 * inside markdown fences, and epic #4435 is extracting that shell into standalone
 * `.sh` files — so the exact bytes this guard used to scan walked off its surface,
 * one merge at a time, while the required CI check kept exiting 0 (#4496, the #4470
 * class). A `.sh` is scanned **whole**, comments included: unlike an *invocation*
 * guard, a machine-local path is a leak wherever it sits in a committed file, and a
 * shell comment is the likeliest place for one to land.
 *
 * The machine-local-path shapes themselves — the generic deny-list design and the
 * `~/.claude` config-file carve-out (#3475/#3505); the `/tmp` arm carries no carve-out,
 * so it fail-closes on any bare `/tmp/…` (#3492 Option 1) — live in the single shared
 * `path-matcher.ts` module this doc/comment scanner imports. (The second consumer, the
 * crew-plugin sanitizer, left with ADR 0279; the single-definition shape is what keeps
 * the drift class closed — the #3506 root bug.) This module owns only the surface
 * scoping (which files get scanned) and the self-exempt list; the path shapes are
 * imported, not re-declared. See `path-matcher.ts` for the pattern rationale.
 */
import {MACHINE_LOCAL_PATH_PATTERNS, type PathPattern, TEMP_PATH_PATTERNS} from "./path-matcher.ts";

export interface Leak {
	/** The exact substring that matched a leak pattern. */
	readonly matched: string;
	/** Human-readable reason, the report line for this leak class. */
	readonly reason: string;
}

const DOC_SUFFIXES = [".md", ".mdx", ".markdown"] as const;
const DOC_DIRS = ["/.decisions/", "/.patterns/"] as const;

// Extension-scoped like DOC_SUFFIXES, and deliberately NOT directory-scoped: confining the
// shell surface to `claude-plugins/` would rebuild the same narrow surface one level down, so
// the next script extracted anywhere else would walk off it exactly as these did. `.sh` is the
// whole tracked shell corpus today — no `.bash`/`.zsh` file exists — so this list is the corpus,
// not a guess at it.
const SHELL_SUFFIXES = [".sh"] as const;

// Files whose subject IS path hygiene: they must spell the forbidden tokens out as
// patterns, so they are exempt by path suffix. Includes the leak-guard's own source
// (the old package AND this moved tool) and the skills that name the patterns.
//
// Exported for the conformance test alone (fabrika-doc-leak-conformance.test.ts): `.fabrika.jsonc`
// carries the doc-surface half of this list a second time, for a predictor that may not import this
// module (ADR 0238/0273), and that copy is pinned against this one rather than promised in prose.
export const DOC_SELF_EXEMPT = [
	"/packages/leak-guard/src/leak-guard.ts",
	"/packages/leak-guard/src/leak-guard.unit.test.ts",
	"/packages/leak-guard/src/bin.ts",
	"/packages/leak-guard/README.md",
	"/packages/pipeline-cli/src/tools/leak-guard/leak-guard.ts",
	"/packages/pipeline-cli/src/tools/leak-guard/leak-guard.unit.test.ts",
	// The shared machine-local-path matcher spells the forbidden token shapes out as
	// patterns (moved out of leak-guard.ts per #3506), so it is exempt like its host was.
	"/packages/pipeline-cli/src/tools/leak-guard/path-matcher.ts",
	"/packages/pipeline-cli/src/tools/leak-guard/path-matcher.unit.test.ts",
	"/packages/pipeline-cli/src/tools/leak-guard/command.ts",
	// Documents the leak-guard deny-list patterns (the ~/.claude / ~/code/ / /vault
	// example shapes), so it must spell the forbidden tokens out — exempt like the
	// old package's own README already is.
	"/packages/pipeline-cli/README.md",
	"/skills/review-doc/SKILL.md",
	"/skills/triage/SKILL.md",
	"/skills/report/SKILL.md",
	// The triager agent's ## Output privacy rule names ~/code/... / /Users/... as the
	// machine-local shapes it forbids in the return summary (#1956) — illustrative rule
	// text, not real paths, so routine edits must not trip the guard.
	"/agents/triager.md",
	"/skills/report/footer.sh",
	// A verification harness whose whole job is proving a corpus carries none of the
	// forbidden shapes, so it must enumerate them as pattern arms — exempt for the same
	// reason path-matcher.ts is, and unreachable by any narrower fix (#4449).
	"/skills/write-code/scripts/verify-fail-closed.sh",
	// Its Lineage section deliberately names ~/code/... sibling-repo clones (the
	// rebuild provenance), so routine edits to it must not trip the guard.
	"/CLAUDE.md",
] as const;

const normalize = (path: string): string => path.replace(/\\/g, "/");

/** Every leak in `text` for `patterns`, deduped on matched+reason (report order = pattern order). */
const scanPatterns = (text: string, patterns: ReadonlyArray<PathPattern>): ReadonlyArray<Leak> => {
	const seen = new Set<string>();
	const leaks: Leak[] = [];
	for (const {pattern, reason} of patterns) {
		for (const match of text.matchAll(pattern)) {
			const matched = match[0];
			const key = `${matched}\t${reason}`;
			if (seen.has(key)) continue;
			seen.add(key);
			leaks.push({matched, reason});
		}
	}
	return leaks;
};

/** Which shared-artifact surface a path belongs to — the scan scope's one classification axis. */
export type Surface = "doc" | "shell";

/**
 * The surface `path` belongs to, or `null` when it is out of scope entirely. Every scoping
 * decision resolves here: `isSharedArtifact` is this function's nullness, and the scan's
 * per-surface scope tally is this function's answer counted. One classifier, so a surface
 * cannot be scanned by one caller and skipped by another (the #3506 drift shape).
 */
export const surfaceOf = (path: string): Surface | null => {
	const p = normalize(path);
	if (SHELL_SUFFIXES.some((s) => p.endsWith(s))) return "shell";
	if (DOC_SUFFIXES.some((s) => p.endsWith(s))) return "doc";
	// Leading-slash the path so a relative `.decisions/x` matches the same `/.decisions/`
	// token as an absolute one (Write passes absolute; be robust to relative targets).
	const rooted = `/${p.replace(/^\/+/, "")}`;
	return DOC_DIRS.some((d) => rooted.includes(d)) ? "doc" : null;
};

export const isSharedArtifact = (path: string): boolean => surfaceOf(path) !== null;

export const isSelfExempt = (path: string): boolean => {
	// Normalize to a leading slash so a relative `.claude/...` target matches the same
	// suffix as an absolute `/…/.claude/...` one (Write passes absolute, but be robust).
	const p = `/${normalize(path).replace(/^\/+/, "")}`;
	return DOC_SELF_EXEMPT.some((s) => p.endsWith(s));
};

/** Every leak in `text` (deduped on matched+reason), scoped to the file surface. */
export const findLeaks = (filePath: string, text: string): ReadonlyArray<Leak> => {
	if (!filePath || !text) return [];
	if (!isSharedArtifact(filePath)) return [];
	if (isSelfExempt(filePath)) return [];
	return scanPatterns(text, MACHINE_LOCAL_PATH_PATTERNS);
};

/**
 * Every machine-local path leak in a PR/issue COMMENT body — the surface `findLeaks` never
 * reaches (it scopes to doc *files*), the gap that let a `review-*` verdict comment carry a
 * `/private/tmp/.../scratchpad/verdict.md` scratchpad ref and a `@/var/folders/.../tmp.XXXX`
 * mktemp path onto public PRs (#2796/#2822/#2683/#2772). A comment body is UNCONDITIONALLY a
 * shared public artifact, so there is no doc-surface gate and no self-exempt list — it scans
 * the shared `MACHINE_LOCAL_PATH_PATTERNS` AND the stricter comment-body-only `TEMP_PATH_PATTERNS`
 * (which fail-closes on any bare `/tmp/…`, no carve-out — #3492 Option 1). Generic path shapes only (#2393).
 */
export const findCommentLeaks = (text: string): ReadonlyArray<Leak> => {
	if (!text) return [];
	return scanPatterns(text, [...MACHINE_LOCAL_PATH_PATTERNS, ...TEMP_PATH_PATTERNS]);
};
